crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,391 @@
+                    return Ok(shared_cap)
+                }
+                _ => continue,
+                // more caps to come ..
+            }
+        }
+
+        Err(CapabilityNotShared)
+    }
+
+    fn reg_new_ingress_buffer(
+        &mut self,
+        cap: SharedCapability,
+    ) -> Result<mpsc::UnboundedReceiver<BytesMut>, MuxDemuxError> {
+        if let Some(tx) = self.demux.get(&cap) {
+            if !tx.is_closed() {
+                return Err(StreamAlreadyExists)
+            }
+        }
+        let (ingress_tx, ingress) = mpsc::unbounded_channel();
+        self.demux.insert(cap, ingress_tx);
+
+        Ok(ingress)
+    }
+
+    fn unmask_msg_id(&self, id: &mut u8) -> Result<SharedCapability, MuxDemuxError> {
+        for cap in self.shared_capabilities.iter_caps() {
+            let offset = cap.relative_message_id_offset();
+            let next_offset = offset + cap.num_messages()?;
+            if *id < next_offset {
+                *id -= offset;
+
+                return Ok(cap)
+            }
+        }
+
+        Err(MessageIdOutOfRange(*id))
+    }
+
+    /// Masks message id with offset relative to the message id suffix reserved for capability
+    /// message ids. The p2p stream further masks the message id (todo: mask whole message id at
+    /// once to avoid copying message to mutate id byte or sink BytesMut).
+    fn mask_msg_id(&self, msg: Bytes) -> Bytes {
+        let mut masked_bytes = BytesMut::zeroed(msg.len());
+        masked_bytes[0] = msg[0] + self.owner.relative_message_id_offset();
+        masked_bytes[1..].copy_from_slice(&msg[1..]);
+
+        masked_bytes.freeze()
+    }
+
+    /// Checks if all clones of this shared stream have been dropped, if true then returns //
+    /// function to drop the stream.
+    fn can_drop(&mut self) -> bool {
+        for tx in self.demux.values() {
+            if !tx.is_closed() {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl<S, E> Stream for MuxDemuxStream<S>
+where
+    S: Stream<Item = Result<BytesMut, E>> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E> + From<<S as Sink<Bytes>>::Error>,
+{
+    type Item = Result<BytesMut, MuxDemuxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // sink buffered bytes from `StreamClone`s
+        for _ in 0..self.demux.len() {
+            let Ok(item) = self.mux.try_recv() else { break };
+            if self.poll_ready_unpin(cx).is_pending() {
+                break
+            }
+            self.inner.start_send_unpin(item)?;
+        }
+
+        // advances the wire and either yields or delegates the message
+        //
+        // poll once for main stream and each stream clone, since stream clones are weak and
+        // cannot poll the `MuxDemuxer` themselves
+        for _ in 0..self.demux.len() + 1 {
```

Comment:
this could result in the stream not being woken up again because we can end up returning Pending if we received `len + 1` messages that don't belong to the owned stream.
so I think we should just use a `loop` here